### ISSUE TITLE
Streaming sessions are not closed cleanly but hang when node stops due to IO error

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#1899]
*Installation details*
Scylla version (or git commit hash): 1.5-rc1
Cluster size: 4

Streaming peer:

```
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 0] storage_service - Stop transport: stop_gossiping done
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 0] storage_service - Thrift server stopped
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 0] storage_service - CQL server stopped
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 0] storage_service - Stop transport: shutdown rpc and cql server done
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 3] rpc - client 172.30.0.161: server connection dropped: Semaphore broken
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 11] rpc - client 172.30.0.161: server connection dropped: Semaphore broken
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 5] rpc - client 172.30.0.161: server connection dropped: Semaphore broken
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 3] rpc - client 172.30.0.163: server connection dropped: Semaphore broken
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 0] storage_service - messaging_service stopped
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 0] storage_service - Stop transport: shutdown messaging_service done
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 11] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] Session with 172.30.0.160 is complete, state=FAILED
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 11] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] bytes_sent = 0, bytes_received = 234850302730
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 0] storage_service - stream_manager stopped
[54.162.11.147] [stdout] Nov 25 11:01:06 warning|  [shard 11] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] Stream failed, peers={172.30.0.160}
[54.162.11.147] [stdout] Nov 25 11:01:06 info   |  [shard 0] storage_service - Stop transport: shutdown stream_manager done
[54.162.11.147] [stdout] Nov 25 11:01:07 info   |  [shard 0] storage_service - Stop transport: auth shutdown
[54.162.11.147] [stdout] Nov 25 11:01:07 info   |  [shard 0] storage_service - Stop transport: done
```

Streaming initiator is not notified about the failure:

```
[54.91.212.220] [stdout] Nov 25 05:09:37 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] Starting streaming to 172.30.0.163
[54.91.212.220] [stdout] Nov 25 05:09:37 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] Beginning stream session with 172.30.0.163
[54.91.212.220] [stdout] Nov 25 05:09:37 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] Prepare completed with 172.30.0.163. Receiving 0, sending 4
[54.91.212.220] [stdout] Nov 25 05:09:37 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] Prepare completed with 172.30.0.159. Receiving 0, sending 4
[54.91.212.220] [stdout] Nov 25 05:09:37 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] Prepare completed with 172.30.0.161. Receiving 0, sending 4
[54.91.212.220] [stdout] Nov 25 11:02:41 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] The session 0x60010aae2c00 made no progress with peer 172.30.0.159
[54.91.212.220] [stdout] Nov 25 11:03:41 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] The session 0x60010aae2c00 made no progress with peer 172.30.0.159
[54.91.212.220] [stdout] Nov 25 11:04:41 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] The session 0x60010aae2c00 made no progress with peer 172.30.0.159
[54.91.212.220] [stdout] Nov 25 11:05:36 warning|  [shard 5] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] stream_transfer_task: Fail to send STREAM_MUTATION to 172.30.0.159:11: rpc::closed_error (connection is closed)
[54.91.212.220] [stdout] Nov 25 11:05:41 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] The session 0x60010aae2c00 made no progress with peer 172.30.0.159
[54.91.212.220] [stdout] Nov 25 11:06:41 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] The session 0x60010aae2c00 made no progress with peer 172.30.0.159
[54.91.212.220] [stdout] Nov 25 11:07:41 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] The session 0x60010aae2c00 made no progress with peer 172.30.0.159
[54.91.212.220] [stdout] Nov 25 11:08:41 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] The session 0x60010aae2c00 made no progress with peer 172.30.0.159
[54.91.212.220] [stdout] Nov 25 11:09:41 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] The session 0x60010aae2c00 made no progress with peer 172.30.0.159
[54.91.212.220] [stdout] Nov 25 11:10:41 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] The session 0x60010aae2c00 made no progress with peer 172.30.0.159
[54.91.212.220] [stdout] Nov 25 11:11:41 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] The session 0x60010aae2c00 is idle for 600 seconds, the peer 172.30.0.159 is probably gone, close it
[54.91.212.220] [stdout] Nov 25 11:11:41 warning|  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] Streaming error occurred
[54.91.212.220] [stdout] Nov 25 11:11:41 info   |  [shard 0] stream_session - [Stream #5cfea190-b2cd-11e6-bcc7-000000000007] Session with 172.30.0.159 is complete, state=FAILED
```

**Comment by asias** [comment]
I think the issue is fixed by the following patch already. @slivne @tgrabiec 

With it,  the failed session will notify peer to close the session.

```
commit aa87429e67103e9b862b7a0a4e7020384bc434c2
Author: Asias He <asias@scylladb.com>
Date:   Tue Jul 18 15:44:34 2017 +0800

    streaming: Send complete message with failed flag when session is failed

    To notify peer node the session is failed.
```

**Comment by asias** [comment]
@slivne ping

**Comment by slivne** [comment]
Fixed in 2.1
